Add devenv based devcontainer for GH Codespaces

datakurre opened the pull request:
@fredvd May I add GH Codespace support for @rioksane to easier play with the and propose changes. This does not really touch any existing parts and should be pretty harmless.

I acknowledge that this is based on use of Nix / Devenv, but I hope it does not matter as long as it does not break existing use.

Example: https://github.com/codespaces/new/datakurre/2025.ploneconf.org (it does take a few minutes to build)

fredvd commented:
YES, PLEASE!  I'd love to see more examples of devcontainer usage & Plone and how we can provide them. 

If you want we can do a short meeting to discuss what else you would like.  But basically: everything that we (Erico/me/sysadmin-team) have provided for the ploneconf website is to give you a head start, and to make sure we can keep the ploneconf websites online, access them and maintained at least a few years after the conference, by standardizing so required time is minimal.  Anything you can add, show case or want in the project,  let us know. 

The conf websites are one of the public show cases where i often refer others too, if they want to know how something works (worked out less).  It's also public documentation.

fredvd reviewed (approved):
LGTM for a first start to get working on this. And as I wrote, internally it is also your project ;-) 

We discussed the importance of devcontainers last year at the conf. But limited resources so far. I know @davisagli also spent some time on getting devcontainers to work, checking issues like port forwarding etc. 

One comment I can make is the .pre-commit-config.yaml, as in other efforts we'd like to stay clear from pre-commit for reasons.  @ericof  ?  And it would be nice if we can keep the top level directory as clean as possible.

I've been working with uv and hatchling myself to get another project upgraded in the last two weeks, with the latest cookieplone improvements. 

These go in the direction that @ericof and others likely are working on this week at the Bucharest Sprint.  The /backend/overrrides.nix folder concerns me a bit as there is a lot of rewriting back to setuptools and other low level stuff, that I only have a slight clue about, but seems like hard earned developer wisdom I'd like to know more of ;-)

datakurre commented:
@fredvd Thanks for noticing pre-commit. I’ll remove that and add to ignore before merge.

datakurre commented:
@fredvd Thanks for the review. I removed the accidentally committed `.pre-commit-config.yaml` and added it into `.gitignore`.

More on the setup I'm trying to sneak in:

TL; DR; With Nix / Devenv the DevContainer practice is to use common minimal basic image with just Nix / Devenv (without building custom image), and all the other dependencies are fetched on the launch of the DevContainer. With proper setup, use of Nix binary caches should prevent any compilation and only require download of dependencies – a bit like fetching a lot of layers for huge container images.

Devenv. I've been so long in the Nix rabbit-hole that I am unable to see solutions without it. For me it provides single configuration and packaging language on top of polyglot projects, and multi-process setups without knowing language specific container ecosystems. Devenv.sh, on the other hand, is a project by Plone-alumni Domen Kožar, which tries to hide Nix behind easier developer tooling and documentation.

 `./devenv.yaml`, `./devenv.nix` and `./devenv.lock`. Are devenv configuration files. Those are always required with Devenv.sh.
 
 `./.treefmt.toml` is just a devenv popularized practice to format code with https://treefmt.com/, but because I it is only used for Nix-files, maybe I'll remove it after writing this comment
 
 `./frontend/devenv.nix` configures Node.js and PNPM for building the frontend. It could be merged with `./devenv.nix`, but splitting frontend and backend into separate files should be best practice
 
 `./backend/devenv.nix` is a bit more complex, because I added full Nix-build for an uv locked envronment. Similar minimal setup to `./frontend/devenv.nix` might work as well, but I have not tried it. Fred is right that `./backend/overrides.nix` might broke if packages are updated and their packaging changes in newer versions. `./backend/overrides.nix` is required to define Python packaging tooling when current Nix-tooling was unable to resolve it. I also added po-compilation for packages missing compiled files, because resulting Nix-build is immutable.
 
I'll see, if the simpler less Nix'd `./backend` works. It would be less scary, and could probably use those awesome make files.

datakurre commented:
Ok. I believe, I can simplify this setup a lot to make it simpler (yet less immutable, which is not necessarily a bad thing here). I'll comment when ready.

datakurre commented on .vscode/settings.json:
```diff
@@ -11,5 +11,6 @@
     "editor.formatOnSave": false
-  }
+  },
+  "makefile.configureOnOpen": false
```
This is automatically added by GH Codespaces.

datakurre commented:
Ok. This is much simpler now. Also tried a couple of times that it should also work. Updated Makefile target to `make codespace-start`.

fredvd commented:
Merged, and I've given you access to the repo @datakurre
